subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/LazyTaskExecution.java

```diff
@@ -0,0 +1,205 @@
+        );
+        this.snapshotRepository = snapshotRepository;
+    }
+
+    @Override
+    public ImmutableSortedMap<String, FileCollectionSnapshot> getInputFilesSnapshot() {
+        if (inputFilesSnapshot == null) {
+            ImmutableSortedMap.Builder<String, FileCollectionSnapshot> builder = ImmutableSortedMap.naturalOrder();
+            for (Map.Entry<String, Long> entry : inputFilesSnapshotIds.entrySet()) {
+                builder.put(entry.getKey(), snapshotRepository.get(entry.getValue()));
+            }
+            inputFilesSnapshot = builder.build();
+        }
+        return inputFilesSnapshot;
+    }
+
+    @Override
+    public FileCollectionSnapshot getDiscoveredInputFilesSnapshot() {
+        if (discoveredFilesSnapshot == null) {
+            discoveredFilesSnapshot = snapshotRepository.get(discoveredFilesSnapshotId);
+        }
+        return discoveredFilesSnapshot;
+    }
+
+    @Override
+    public void setDiscoveredInputFilesSnapshot(FileCollectionSnapshot discoveredFilesSnapshot) {
+        this.discoveredFilesSnapshot = discoveredFilesSnapshot;
+        this.discoveredFilesSnapshotId = null;
+    }
+
+    @Override
+    public ImmutableSortedMap<String, FileCollectionSnapshot> getOutputFilesSnapshot() {
+        if (outputFilesSnapshot == null) {
+            ImmutableSortedMap.Builder<String, FileCollectionSnapshot> builder = ImmutableSortedMap.naturalOrder();
+            for (Map.Entry<String, Long> entry : outputFilesSnapshotIds.entrySet()) {
+                String propertyName = entry.getKey();
+                builder.put(propertyName, snapshotRepository.get(entry.getValue()));
+            }
+            outputFilesSnapshot = builder.build();
+        }
+        return outputFilesSnapshot;
+    }
+
+    @Override
+    public void setOutputFilesSnapshot(ImmutableSortedMap<String, FileCollectionSnapshot> outputFilesSnapshot) {
+        this.outputFilesSnapshot = outputFilesSnapshot;
+        this.outputFilesSnapshotIds = null;
+    }
+
+    public TaskExecutionSnapshot snapshot() {
+        return new TaskExecutionSnapshot(
+            isSuccessful(),
+            getBuildInvocationId(),
+            getTaskImplementation(),
+            getTaskActionImplementations(),
+            getOutputPropertyNamesForCacheKey(),
+            getDeclaredOutputFilePaths(),
+            getInputProperties(),
+            inputFilesSnapshotIds,
+            discoveredFilesSnapshotId,
+            outputFilesSnapshotIds
+        );
+    }
+
+    public void storeSnapshots() {
+        if (inputFilesSnapshotIds == null && inputFilesSnapshot != null) {
+            ImmutableSortedMap.Builder<String, Long> builder = ImmutableSortedMap.naturalOrder();
+            for (Map.Entry<String, FileCollectionSnapshot> entry : inputFilesSnapshot.entrySet()) {
+                builder.put(entry.getKey(), snapshotRepository.add(entry.getValue()));
+            }
+            inputFilesSnapshotIds = builder.build();
+        }
```

Comment:
General | Duplicated Code. See https://www.jetbrains.com/help/idea/analyzing-duplicates.html (on the fly analysis)